aten/src/ATen/Parallel_native.h

```diff
@@ -0,0 +1,181 @@
+#pragma once
+#include <ATen/ATen.h>
+#include <ATen/core/ivalue.h>
+#include <ATen/Parallel.h>
+
+#include <c10/core/thread_pool.h>
+
+#include <algorithm>
+#include <cstddef>
+#include <exception>
+
+#define INTRA_OP_PARALLEL
+
+namespace at {
+namespace internal {
+// internal function to get access to intra-op thread pool from
+// template parallel primitives (parallel_for, parallel_reduce)
+CAFFE2_API TaskThreadPoolBase& _get_intraop_pool();
+
+// internal utility function to mark master thread as in parallel
+// region when executing parallel primitives
+CAFFE2_API void _set_in_parallel_region(bool);
+
+// Simulate OMP's omp_get_thread_num() by force-setting thread local
+// task id as thread number when executing parallel primitives
+CAFFE2_API void _set_thread_num(size_t thread_num);
+CAFFE2_API void _unset_thread_num();
+}
+
+template <class F>
```

Comment:
Can you add comments describing what these do at a high level? If these are to be used in operator implementations, we should have some really good docs so people know what to do e.g. for setting `grain_size`

Reply:
Will add, I'd suggest only to add this into generic interface file (Parallel.h) - we already have a good comment on `parallel_reduce`, need to add some for `parallel_for` there too